transport/peer/x/peerlist/roundrobin/roundrobin.go

```diff
@@ -82,51 +110,54 @@ func (pl *RoundRobin) clearPeers() error {
 
 	peers := pl.pr.RemoveAll()
 	for _, p := range peers {
-		err := pl.agent.ReleasePeer(p, pl)
-		if err != nil {
+		if err := pl.agent.ReleasePeer(p, pl); err != nil {
 			errs = append(errs, err)
 		}
 	}
 
 	return yerrors.MultiError(errs)
 }
 
+// Remove a peer identifier from the round robin
+func (pl *RoundRobin) Remove(pid transport.PeerIdentifier) error {
+	if err := pl.pr.Remove(pid); err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.agent.ReleasePeer(pid, pl)
+}
+
 // ChoosePeer selects the next available peer in the round robin
-func (pl *RoundRobin) ChoosePeer(context.Context, *transport.Request) (transport.Peer, error) {
+func (pl *RoundRobin) ChoosePeer(ctx context.Context, req *transport.Request) (transport.Peer, error) {
 	if !pl.started.Load() {
 		return nil, errors.ErrPeerListNotStarted("RoundRobinList")
 	}
 
-	nextPeer := pl.pr.Next()
-	if nextPeer == nil {
-		return nil, errors.ErrNoPeerToSelect("RoundRobinList")
-	}
-	return nextPeer, nil
-}
-
-// Add a peer identifier to the round robin
-func (pl *RoundRobin) Add(pid transport.PeerIdentifier) error {
-	return pl.addPeer(pid)
-}
+	for {
+		if nextPeer := pl.pr.Next(); nextPeer != nil {
+			return nextPeer, nil
+		}
 
-func (pl *RoundRobin) addPeer(pid transport.PeerIdentifier) error {
-	p, err := pl.agent.RetainPeer(pid, pl)
-	if err != nil {
-		return err
+		if err := pl.waitForPeerAddedEvent(ctx); err != nil {
+			return nil, err
+		}
 	}
-
-	return pl.pr.Add(p)
 }
 
-// Remove a peer identifier from the round robin
-func (pl *RoundRobin) Remove(pid transport.PeerIdentifier) error {
-	err := pl.pr.Remove(pid)
-	if err != nil {
-		// The peer has already been removed
-		return err
+// waitForPeerAddedEvent waits until a peer is added to the peer list or the
+// given context finishes.
+func (pl *RoundRobin) waitForPeerAddedEvent(ctx context.Context) error {
+	if _, ok := ctx.Deadline(); !ok {
+		return errors.ErrChooseContextHasNoDeadline("RoundRobinList")
 	}
 
-	return pl.agent.ReleasePeer(pid, pl)
+	select {
+	case <-pl.peerAddedEvent:
```

Comment:
waaaaaaaaaaaaait, won't this only block for the *first* peer added event?  If we get a bunch of requests blocking on the PeerAdded event, only one will get popped off.